index.d.ts

```diff
@@ -1,57 +1,93 @@
-import { type EventEmitter } from 'events'
+import events from 'events'
+import amqplib from 'amqplib'
 
 export as namespace hato;
 
 /**
- * Plugins
+ * Types
  */
-export class Plugin {
-    constructor(name?: string)
-}
-
-export class RPC extends Plugin {
-    constructor({ timeout }: { timeout: number })
-}
+export type ExchangeType = 'fanout' | 'direct' | 'topic' | 'headers';
 
-export class Retry extends Plugin {
-    constructor({ base }: { base: number })
-}
+export interface Consumer extends Promise<amqplib.Replies.Consume>, events.EventEmitter {}
 
-export class ServiceContext extends Plugin {
-    constructor({ name, version }: { name: string, version:  string })
+export interface ConsumeMessage extends amqplib.ConsumeMessage {
+    ack(allUpTo?: boolean): void
+    nack(allUpTo?: boolean, requeue?: boolean): void
+    cancel(): void
 }
 
-type Defaults = {
-    prefetch: number
-    queue: {
-        durable: boolean
-    },
-    exchange: {
-        durable: boolean
+/**
+ * Plugins
+ */
+export namespace Plugins {
+    abstract class Base {
+        constructor(name: string)
+        abstract init(): void;
+        abstract destroy(): void;
     }
-    publish: {
-        persistent: boolean
-        mandatory: boolean
+    interface Plugin<T> extends Plugins.Base {
+        new(options?: T): Plugin<T>
+    }
+    interface Defaults {
+        Base: Base
+        GracefulShutdown: Plugin<Options.GracefulShutdown>
+        ConnectionRetry: Plugin<Options.ConnectionRetry>
+        Duplex: Plugin<Options.Duplex>
+        DLX: Plugin<Options.DLX>
+        Encoding: Plugin<Options.Encoding>
+        RPC: Plugin<Options.RPC>
+        Confirm: Plugin<Options.Confirm>
+        Retry: Plugin<Options.Retry>
+        ServiceContext: Plugin<Options.ServiceContext>
+        DefaultOptions: Plugin<Options.DefaultOptions>
+    }
+    namespace Options {
+        type GracefulShutdown = null;
```

Comment:
I just didn't get this part, could you explain why you made a type instead of using null directly?

Reply:
I totally could (have done `Plugin<null>` above). I just thought it would be nice all the plugin option types are available in the namespace, like for a documentation purpose

Reply:
Nice, good idea